matchbox_signaling/src/signaling_server/handlers.rs

```diff
@@ -0,0 +1,114 @@
+use crate::{
+    signaling_server::{auth::AuthKey, callbacks::SharedCallbacks, Authentication, SignalingState},
+    topologies::{
+        common_logic::{spawn_sender_task, try_send, SignalingChannel},
+        SignalingStateMachine,
+    },
+    SignalingCallbacks,
+};
+use axum::{
+    extract::{
+        ws::{Message, WebSocket},
+        ConnectInfo, Path, Query, WebSocketUpgrade,
+    },
+    response::IntoResponse,
+    Extension,
+};
+use futures::{stream::SplitStream, StreamExt};
+use hyper::{HeaderMap, StatusCode};
+use matchbox_protocol::{JsonPeerEvent, PeerId};
+use std::{collections::HashMap, net::SocketAddr};
+use tracing::{error, info};
+
+/// Metastate used during by a signaling server's runtime
+pub struct WsStateMeta<Cb, S> {
+    pub peer_id: PeerId,
+    pub sender: SignalingChannel,
+    pub receiver: SplitStream<WebSocket>,
+    pub callbacks: Cb,
+    pub state: S,
+}
+
+/// Metadata captured at the time of websocket upgrade
+#[derive(Debug, Clone)]
+pub struct WsUpgradeMeta {
+    pub origin: SocketAddr,
+    pub path: Option<String>,
+    pub query_params: HashMap<String, String>,
+    pub headers: HeaderMap,
+}
+
+/// The handler for the HTTP request to upgrade to WebSockets.
+/// This is the last point where we can extract metadata such as IP address of the client.
+#[allow(clippy::too_many_arguments)]
+pub(crate) async fn ws_handler<Cb, S, A>(
+    ws: WebSocketUpgrade,
+    path: Option<Path<String>>,
+    headers: HeaderMap,
+    Query(query_params): Query<HashMap<String, String>>,
+    Extension(shared_callbacks): Extension<SharedCallbacks>,
+    Extension(callbacks): Extension<Cb>,
+    Extension(auth_key): Extension<AuthKey>,
+    Extension(state): Extension<S>,
+    Extension(state_machine): Extension<SignalingStateMachine<Cb, S>>,
+    ConnectInfo(origin): ConnectInfo<SocketAddr>,
+) -> impl IntoResponse
+where
+    Cb: SignalingCallbacks,
+    S: SignalingState,
+    A: Authentication,
+{
+    info!("`{origin}` connected.");
+
+    let path = path.map(|path| path.0);
+    let meta = WsUpgradeMeta {
+        origin,
+        path,
+        query_params,
+        headers,
+    };
+
+    // Lifecycle event: On Connection Request
+    match shared_callbacks.on_connection_request.emit(meta.clone()) {
+        Ok(true) => {}
+        Ok(false) => return (StatusCode::UNAUTHORIZED).into_response(),
+        Err(e) => return e,
+    };
+
+    // Check authentication
+    match A::verify(meta, auth_key) {
+        Ok(true) => {}
+        Ok(false) => return (StatusCode::UNAUTHORIZED).into_response(),
+        Err(e) => return e,
+    };
+
+    // Finalize the upgrade process by returning upgrade callback to client
+    // Generate an ID for the peer
+    let peer_id = uuid::Uuid::new_v4().into();
```

Comment:
There was some discussion of making it possible to override this (e.g. on reconnections/meaningful uuids), was this left for a follow up PR?

Reply:
There's callbacks for id assignment, but no way to not send an ID. That can be future work if someone wants it.